metagraph/src/graph/alignment/aligner_helper.hpp

```diff
@@ -515,10 +520,71 @@ class DPTable {
         char last_char;
         size_t best_pos;
         size_t last_priority_pos;
+        size_t start_index;
+
+        void expand_to_cover(size_t begin, size_t end) {
+            assert(best_pos >= start_index);
+            assert(best_pos - start_index < scores.size());
+            assert(last_priority_pos >= start_index);
+            assert(last_priority_pos - start_index < scores.size());
+
+            if (begin >= start_index) {
+                // the current range already covers [begin, end)
+                if (end <= start_index + scores.size() - 8)
+                    return;
+
+                // extend the range to the right to reach end
+                scores.resize(end + 8 - start_index);
+                gap_scores.resize(end + 8 - start_index);
+                ops.resize(end + 8 - start_index);
+                prev_nodes.resize(end + 8 - start_index);
+                gap_prev_nodes.resize(end + 8 - start_index);
+                gap_count.resize(end + 8 - start_index);
+            } else if (end <= start_index + scores.size() - 8) {
+                // extend the range to the left to reach begin
+                size_t shift = start_index - begin;
+                start_index = begin;
+                scores.insert(scores.begin(), shift, min_score_);
+                gap_scores.insert(gap_scores.begin(), shift, min_score_);
+                ops.insert(ops.begin(), shift, Cigar::CLIPPED);
+                prev_nodes.insert(prev_nodes.begin(), shift, 0);
+                gap_prev_nodes.insert(gap_prev_nodes.begin(), shift, 0);
+                gap_count.insert(gap_count.begin(), shift, 0);
+            } else {
+                // extend the range in both directions
+                size_t shift = start_index - begin;
+                start_index = begin;
+                end += 8;
+                size_t new_size = end - begin;
+                AlignedVector<score_t> new_scores(new_size, min_score_);
+                AlignedVector<score_t> new_gap_scores(new_size, min_score_);
+                AlignedVector<Cigar::Operator> new_ops(new_size, Cigar::CLIPPED);
+                AlignedVector<uint8_t> new_prev_nodes(new_size);
+                AlignedVector<uint8_t> new_gap_prev_nodes(new_size);
+                AlignedVector<int32_t> new_gap_count(new_size);
+                std::copy(scores.begin(), scores.end(), new_scores.begin() + shift);
+                std::copy(gap_scores.begin(), gap_scores.end(), new_gap_scores.begin() + shift);
+                std::copy(ops.begin(), ops.end(), new_ops.begin() + shift);
+                std::copy(prev_nodes.begin(), prev_nodes.end(), new_prev_nodes.begin() + shift);
+                std::copy(gap_prev_nodes.begin(), gap_prev_nodes.end(), new_gap_prev_nodes.begin() + shift);
+                std::copy(gap_count.begin(), gap_count.end(), new_gap_count.begin() + shift);
+                std::swap(scores, new_scores);
+                std::swap(gap_scores, new_gap_scores);
```

Comment:
for each
```c++
AlignedVector<score_t> new_gap_scores(new_size, min_score_);
std::copy(gap_scores.begin(), gap_scores.end(), new_gap_scores.begin() + shift);
std::swap(gap_scores, new_gap_scores);
```

better do
reserve + insert(begin, ...) + insert(end, ...)

It will look simpler